### PR TITLE
Story2 staff register

### PR DESCRIPTION
**Pull request by karimkefi**
story2 Register Staff

### CODE REVIEW

File: public/js/registerUser.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+let newUserForm = document.getElementById('registerUserForm')
+newUserForm.addEventListener('submit', async (e) => {
+    e.preventDefault()
+    let data = getInputs(),
+        messageBox = document.getElementById('message'),
+        button = document.getElementById('newUserSubmit'),
+        response = await sendNewUserDetails(data)
+
+    if(!response['success']) {
+        messageBox.innerText = 'Error adding user to database'
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply public/js/registerUser.js]: Resolved.

[the remark this answers is not in the document]
Comment (marty-crane) [reply public/js/registerUser.js]: resolved

File: public/js/registerUser.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        messageBox.classList.add('error')
+    } else if (response['success'] === true) {
+        messageBox.innerText = `New user added email:' ${data['email']} ' password:' ${data['password']} '`
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply public/js/registerUser.js]: resolved

File: src/Classes/Controllers/RegisterUserController.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    function __invoke(Request $request, Response $response){
+
+        if ($_SESSION['loggedIn'] === true) {
+
+            $data = ['success' => false, 'msg' => 'User not registered.', 'data' => []];
+            $statusCode = 401;
+
+            $newUserData = $request->getParsedBody();
+            $validatedUserData = [
+                'email' => filter_var($newUserData['email'], FILTER_SANITIZE_STRING),
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply src/Classes/Controllers/RegisterUserController.php]: resolved

File: src/Classes/Factories/AdminControllerFactory.php [hunk 4]
```diff
@@ -11,4 +11,7 @@ public function __invoke(ContainerInterface $container) : AdminController {
         $renderer = $container->get('renderer');
         return new AdminController($renderer);
     }
-}
\ No newline at end of file
+}
+
+// create RegisterUserControllerFactory
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply src/Classes/Factories/AdminControllerFactory.php]: removed

File: src/Classes/Controllers/RegisterUserController.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * Validates username / email and updates database if email doesn't exist
+     *
+     * @param Request $request HTTP request
+     * @param Response $response HTTP response
+     *
+     * @return error/success message and status code
+     */
+    function __invoke(Request $request, Response $response){
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply src/Classes/Controllers/RegisterUserController.php]: resolved

File: src/Classes/Models/RandomPasswordModel.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+        //return the transfer as a string
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+
+
+        // $output contains the output string
+        $password = json_decode(curl_exec($ch));
+        shuffle($password);
+
+        //
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply src/Classes/Models/RandomPasswordModel.php]: removed

File: src/Classes/Models/RandomPasswordModel.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        $password = array_pop($password)->password;
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply src/Classes/Models/RandomPasswordModel.php]: resolved

File: src/Classes/Models/RandomPasswordModel.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+        // close curl resource to free up system resources
+        curl_close($ch);
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply src/Classes/Models/RandomPasswordModel.php]: resolved

File: tests/Models/RandomPasswordModelTest.php [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+namespace Tests\Models;
+
+use PHPUnit\Framework\TestCase;
+use Portal\Models\RandomPasswordModel;
+
+class RandomPasswordModelTest extends TestCase
+{
+    function testConstruct(){
+        $password = new RandomPasswordModel();
+        $case = gettype($password());
```
[the remark this answers is not in the document]
Comment (karimkefi) [reply tests/Models/RandomPasswordModelTest.php]: Test removed and commented out

File: src/Classes/Models/UserModel.php [hunk 10]
```diff
@@ -36,9 +36,9 @@ public function getUserByEmail($userEmail)
      * @param string $password value provided for comparison
      * @param array $userCredentials values provided for comparison
      *
-     * @return bool
+     * @return true if email entered exists in database
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply src/Classes/Models/UserModel.php]: resolved

File: tests/Models/RandomPasswordModelTest.php [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Tests\Models;
+
+use PHPUnit\Framework\TestCase;
+use Portal\Models\RandomPasswordModel;
+
+class RandomPasswordModelTest extends TestCase
+{
+    function testConstruct(){
```
[the remark this answers is not in the document]
Comment (karimkefi) [reply tests/Models/RandomPasswordModelTest.php]: test removed (see later comment)

File: tests/Factories/RegisterControllerFactoryTest.php [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+class RegisterControllerFactoryTest extends TestCase
+{
+    function testInvoke()
+    {
+        $container = $this->createMock(ContainerInterface::class);
+        $renderer = $this->createMock(PhpRenderer::class);
+        $RandomPasswordModel = $this->createMock(RandomPasswordModel::class);
+
+        $container->expects($this->at(0))->method('get')//best solution is to use prophecy but this works. Do not mess with order in factory
```
[the remark this answers is not in the document]
Comment (marty-crane) [reply tests/Factories/RegisterControllerFactoryTest.php]: resolved

File: tests/Models/RandomPasswordModelTest.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (marty-crane) [reply tests/Models/RandomPasswordModelTest.php]: resolved

File: tests/Models/RandomPasswordModelTest.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (marty-crane) [reply tests/Models/RandomPasswordModelTest.php]: resolved

**Pull request merged**